lib/libutee/tee_api.c

```diff
@@ -404,12 +420,16 @@ void *TEE_Malloc(uint32_t len, uint32_t hint)
 
 void *TEE_Realloc(void *buffer, uint32_t newSize)
 {
-	return realloc(buffer, newSize);
```

Comment:
I was expected something like this:
``` C
if (is_zero_sized(buffer)) 
        return TEE_Malloc(newSize, get_zero_sized_hints(buffer));

return realloc(buffer, newSize);

Reply:
I don't want to derail the review, but... how does this handle the hint?
```
return realloc(buffer, newSize);
```

Reply:
It currently doesn't. I thought we were storing the hint somewhere, but as it turns out `realloc()` is always clearing the remaining part of a buffer.

With that in mind, perhaps I'm asking for a bit much.

Reply:
The hint is the address itself. Something like:

```c
void *TEE_Realloc(void *buffer, uint32_t newSize)
{
	if (!newSize) {
		free(buffer);
		return (void *)1;
	}

	if (buffer == (void *)1)
		return malloc(newSize);

	return realloc(buffer, newSize);
}
```

Reply:
Hah, it seems we all suggest half broken solutions. :-)
If the original zero sized buffer was obtained with `TEE_Malloc(0, TEE_MALLOC_FILL_ZERO)`, then the caller would be right to expect a zeroed buffer. So `calloc()` would be a better option than malloc() here.

Reply:
you're right:
```patch
+ #define NULLSIZE_VA ((void *)1)
+
void *TEE_Malloc(uint32_t len, uint32_t hint)
{
+	if (!len)
+		return NULLSIZE_VA;
+
	if (hint == TEE_MALLOC_FILL_ZERO)
		return calloc(1, len);
	else if (hint == TEE_USER_MEM_HINT_NO_FILL_ZERO)
		return malloc(len);
	EMSG("Invalid hint %#" PRIx32, hint);
	return NULL;
}

void *TEE_Realloc(void *buffer, uint32_t newSize)
{
+	if (!newSize) {
+		free(buffer);
+		return NULLSIZE_VA;
+	}
+
+	if (buffer == NULLSIZE_VA)
+		return calloc(1, newSize);
+
	return realloc(buffer, newSize);
}

void TEE_Free(void *buffer)
{
-	free(buffer);
+	if (buffer != NULLSIZE_VA)
+		free(buffer);
}

Reply:
Replace the call to `free()` with `TEE_Free()` in `TEE_Realloc()` and we should be done.

Reply:
right :)
@Sadiq1811, can you update your P-R with the outcome of the discussion?

Reply:
done